File: IPWorks Samples/HTML Mailer/javascript/htmlmailer.js
/*
 * IPWorks 2022 JavaScript Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
 
const readline = require("readline");
const ipworks = require("@nsoftware/ipworks");

if(!ipworks) {
  console.error("Cannot find ipworks.");
  process.exit(1);
}
let rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

main();

async function main() {

	const argv = process.argv;
	if(argv.length < 9 || argv.length > 16) {
		console.log("Usage: node htmlmailer.js [-option parameter] server from to");
		console.log("Options: ");
		console.log("  -ssl    a switch parameter to enable SSL");
		console.log("  -u      the username used to authenticate to the mail server");
		console.log("  -p      the password of the user");		
		console.log("  -s      the subject of the mail message");
		console.log("  -m      the raw message content");
		console.log("  -a      the path of file to attach to the message");
		console.log("  server  the name or address of a mail server");			
		console.log("  from    the email address of the sender");
		console.log("  to      a comma separated list of addresses for destinations");
		console.log("Example: node htmlmailer.js -ssl -u test@mydomain.com -p mypassword -s test -m \"<b>Hello</b>, my name is <i>Tom</i>\" -a \"C:\\File\\To\\Attach.jpg\" smtp.gmail.com sender@mail.com recipient1@gmail.com,recipient2@gmail.com");
		process.exit();
	} 

	const htmlmailer1 = new ipworks.htmlmailer();

	htmlmailer1.on("SSLServerAuthentication", function(e) {
		e.accept = true;
	});

	for (i=0; i<argv.length; i++) {
		if (argv[i].startsWith("-")) {
			if (argv[i] === "-ssl") htmlmailer1.setSSLEnabled(true);
			if (argv[i] === "-s")   htmlmailer1.setSubject(argv[i+1]);
			if (argv[i] === "-m")   htmlmailer1.setMessageHTML(argv[i+1]);
			if (argv[i] === "-a")   htmlmailer1.addAttachment(argv[i+1]);
			if (argv[i] === "-u")   htmlmailer1.setUser(argv[i+1]);
			if (argv[i] === "-p")   htmlmailer1.setPassword(argv[i+1]);
		}
	}

	htmlmailer1.setTimeout(60);
	htmlmailer1.setMailServer(argv[argv.length-3]);
	htmlmailer1.setFrom(argv[argv.length-2]);			  
	htmlmailer1.setSendTo(argv[argv.length-1]);


	console.log('Sending message...');

	try {
		await htmlmailer1.send();
		console.log("Message sent successfully.");
	} catch (ex) {
		console.log("Error: " + ex); 
	}
	process.exit();
}

function prompt(promptName, label, punctuation, defaultVal)
{
  lastPrompt = promptName;
  lastDefault = defaultVal;
  process.stdout.write(`${label} [${defaultVal}] ${punctuation} `);
}

File: IPWorks Samples/HTML Mailer/net/htmlmailer.cs
/*
 * IPWorks 2024 .NET Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 * 
 */

﻿using System;
using nsoftware.IPWorks;

class htmlmailerDemo
{
  private static HTMLMailer htmlmailer;

  private static void htmlmailer_OnSSLServerAuthentication(object sender, HTMLMailerSSLServerAuthenticationEventArgs e)
  {
    if (e.Accept) return;
    Console.Write("Server provided the following certificate:\nIssuer: " + e.CertIssuer + "\nSubject: " + e.CertSubject + "\n");
    Console.Write("The following problems have been determined for this certificate: " + e.Status + "\n");
    Console.Write("Would you like to continue anyways? [y/n] ");
    if (Console.Read() == 'y') e.Accept = true;
  }

  static void Main(string[] args)
  {

    htmlmailer = new HTMLMailer();

    if (args.Length < 3)
    {

      Console.WriteLine("usage: htmlmailer [options] server from to");
      Console.WriteLine("Options: ");
      Console.WriteLine("  -s      the subject of the mail message");
      Console.WriteLine("  -m      the HTML version of the message content");
      Console.WriteLine("  -a      the path of file to attach to the message");
      Console.WriteLine("  server  the name or address of a mail server (mail relay)");
      Console.WriteLine("  from    the email address of the sender");
      Console.WriteLine("  to      a comma separated list of addresses for destinations");
      Console.WriteLine("\r\nExample: htmlmailer -s test -m \"<b>Hello</b>, my name is <i>Tom</i>\" -a FileToAttach mail.local sender@mail.com recipient@mail.local");

    }
    else
    {
      try {
        htmlmailer.OnSSLServerAuthentication += htmlmailer_OnSSLServerAuthentication;

        htmlmailer.MailServer = args[args.Length - 3];
        htmlmailer.From = args[args.Length - 2];
        htmlmailer.SendTo = args[args.Length - 1];

        for (int i = 0; i < args.Length; i++)
        {
          if (args[i].StartsWith("-"))
          {
            if (args[i].Equals("-s")) htmlmailer.Subject = args[i + 1]; // args[i+1] corresponds to the value of argument [i]
            if (args[i].Equals("-m")) htmlmailer.MessageHTML = args[i + 1];
            if (args[i].Equals("-a")) htmlmailer.AddAttachment(args[i + 1]); //if you want to add attachment
          }
        }

        //Use these properties for client authentication
        //htmlmailer.setUser(prompt("User"));
        //htmlmailer.setPassword(prompt("Password"));

        Console.WriteLine("Sending message ...");
        htmlmailer.Send();

        Console.WriteLine("Message sent successfully");
      } catch (Exception e)
      {
        Console.WriteLine(e.Message);
      }
      Console.WriteLine("Press any key to exit...");
      Console.ReadKey();
    }
  }
}





class ConsoleDemo
{
  /// <summary>
  /// Takes a list of switch arguments or name-value arguments and turns it into a dictionary.
  /// </summary>
  public static System.Collections.Generic.Dictionary<string, string> ParseArgs(string[] args)
  {
    System.Collections.Generic.Dictionary<string, string> dict = new System.Collections.Generic.Dictionary<string, string>();

    for (int i = 0; i < args.Length; i++)
    {
      // Add a key to the dictionary for each argument.
      if (args[i].StartsWith("/"))
      {
        // If the next argument does NOT start with a "/", then it is a value.
        if (i + 1 < args.Length && !args[i + 1].StartsWith("/"))
        {
          // Save the value and skip the next entry in the list of arguments.
          dict.Add(args[i].ToLower().TrimStart('/'), args[i + 1]);
          i++;
        }
        else
        {
          // If the next argument starts with a "/", then we assume the current one is a switch.
          dict.Add(args[i].ToLower().TrimStart('/'), "");
        }
      }
      else
      {
        // If the argument does not start with a "/", store the argument based on the index.
        dict.Add(i.ToString(), args[i].ToLower());
      }
    }
    return dict;
  }
  /// <summary>
  /// Asks for user input interactively and returns the string response.
  /// </summary>
  public static string Prompt(string prompt, string defaultVal)
  {
    Console.Write(prompt + (defaultVal.Length > 0 ? " [" + defaultVal + "]": "") + ": ");
    string val = Console.ReadLine();
    if (val.Length == 0) val = defaultVal;
    return val;
  }
}